File: app/src/main/java/ru/mrlargha/stemobile/data/LoginRepository.java
package ru.mrlargha.stemobile.data;

import androidx.annotation.Nullable;

import ru.mrlargha.stemobile.data.model.LoginServerReply;

/**
 * Репозиторий авторизации
 */
public class LoginRepository {

    private static volatile LoginRepository instance;

    private final STEDataSource dataSource;
    private LoginServerReply user = null;

    private LoginRepository(STEDataSource dataSource) {
        this.dataSource = dataSource;
    }

    /**
     * Получить экземпляр репоизитория
     *
     * @param dataSource источник данных
     * @return инстанция <code>LoginRepository</code>
     */
    public static LoginRepository getInstance(STEDataSource dataSource) {
        if (instance == null) {
            instance = new LoginRepository(dataSource);
        }
        return instance;
    }

    /**
     * Авторизован ли пользователь
     *
     * @return true - если пользователь авторизован, false - если нет
     */
    public boolean isLoggedIn() {
        return user.getStatus().equals("ok");
    }

    /**
     * Получить токен авторизации
     *
     * @return токен авторизации
     */
    @Nullable
    public String getToken() {
        if (user != null) {
            return user.getSte_token();
        } else {
            return null;
        }
    }

    /**
     * Получить имя пользователя
     *
     * @return имя пользователя
     */
    @Nullable
    public String getName() {
        if (user != null) {
            return user.getName();
        } else {
            return null;
        }
    }

    /**
     * Завершить сеанс (анулировать токен)
     */
    public void logout() {
        dataSource.logout(user.getSte_token());
        user = null;
    }

    private void setLoggedInUser(LoginServerReply user) {
        this.user = user;
    }

    /**
     * Авторизовать пользователя
     *
     * @param username имя пользователя
     * @param password пароль
     * @return ответ сервера <code>LoginServerReply</code>
     */
    public Result<LoginServerReply> login(String username, String password) {
        Result result = dataSource.login(username, password);
        if (result instanceof Result.Success) {
            setLoggedInUser(((Result.Success<LoginServerReply>) result).getData());
        }
        return result;
    }

    /**
     * Зарегистрировать пользователя
     *
     * @param vk_id    vk_id пользователя
     * @param password пароль
     * @return ответ сервера <code>LoginServerReply</code>
     */
    public Result<LoginServerReply> register(String vk_id, String password) {
        Result result = dataSource.register(vk_id, password);
        if (result instanceof Result.Success) {
            setLoggedInUser(((Result.Success<LoginServerReply>) result).getData());
        }
        return result;
    }

    /**
     * Получить информацию о пользователе
     *
     * @param token токен авторизации
     * @return ответ сервера <code>LoginServerReply</code>
     */
    public Result<LoginServerReply> getInfo(String token) {
        Result result = dataSource.getInfo(token);
        if (result instanceof Result.Success) {
            setLoggedInUser(((Result.Success<LoginServerReply>) result).getData());
        }
        return (Result<LoginServerReply>) result;
    }
}

File: app/src/main/java/ru/mrlargha/stemobile/data/LoginRepository.kt
package ru.mrlargha.stemobile.data

import ru.mrlargha.stemobile.data.model.LoginServerReply

class LoginRepository private constructor(private val dataSource: STEDataSource) {
    private var user: LoginServerReply? = null
    val isLoggedIn: Boolean
        get() = user!!.status == "ok"

    val token: String?
        get() = if (user != null) {
            user!!.ste_token
        } else {
            null
        }

    val name: String?
        get() = if (user != null) {
            user!!.name
        } else {
            null
        }

    fun logout() {
        dataSource.logout(user!!.ste_token)
        user = null
    }

    private fun setLoggedInUser(user: LoginServerReply) {
        this.user = user
    }

    fun login(username: String?, password: String?): Result<LoginServerReply> {
        val result = dataSource.login(username, password)
        if (result is Result.Success<*>) {
            setLoggedInUser((result as Result.Success<LoginServerReply>).data)
        }
        return result as Result<LoginServerReply>
    }

    fun register(vk_id: String?, password: String?): Result<LoginServerReply> {
        val result = dataSource.register(vk_id, password)
        if (result is Result.Success<*>) {
            setLoggedInUser((result as Result.Success<LoginServerReply>).data)
        }
        return result as Result<LoginServerReply>
    }

    fun getInfo(token: String?): Result<LoginServerReply> {
        val result = dataSource.getInfo(token)
        if (result is Result.Success<*>) {
            setLoggedInUser((result as Result.Success<LoginServerReply>).data)
        }
        return result as Result<LoginServerReply>
    }

    companion object {
        @Volatile
        private var instance: LoginRepository? = null

        @JvmStatic
        fun getInstance(dataSource: STEDataSource): LoginRepository? {
            if (instance == null) {
                instance = LoginRepository(dataSource)
            }
            return instance
        }
    }
}